core/src/main/scala/quasar/physical/ts/TSDestination.scala

```diff
@@ -108,28 +108,25 @@ final class TSDestination[F[_]: Concurrent: ContextShift: MonadResourceErr] priv
       }
 
       val r = for {
-        tableName <- Resource.liftF[F, String](tableNameF)
+        tableName <- Stream.eval[F, String](tableNameF)
 
-        _ <- Resource.liftF(
+        _ <- Stream.eval(
             Sync[F].delay(
               log.info(s"(re)creating ${config.database}.${tableName} with schema ${columns.show}")))
 
-        p <- client.exec(cc, "tql", blocker)
+        p <- Stream.resource(client.exec(cc, "tql", blocker))
         _ <- Stream(overwriteDdl(tableName, columns) + "exit;")
           .flatMap(s => Stream.emits(s.getBytes("UTF-8")))
           .through(p.stdin)
-          .compile
-          .resource
-          .drain
 
         cmd = loadCommand(tableName)
-        _ <- Resource.liftF(
+        _ <- Stream.eval(
           Sync[F].delay(
             log.info(s"running remote ingest: $cmd")))
 
-        p <- client.exec(cc, cmd, blocker)
+        p <- Stream.resource(client.exec(cc, cmd, blocker))
 
-        ingest = bytes
+        ingest <- bytes
```

Comment:
Think about this `flatMap`. How many elements are in the `Stream` you're binding on?

The answer is "as many as there are chunks". So then the `p.join` later on in this `for`-comprehension doesn't make any sense. This whole block needs to be restructured. Basically, think about what things need to be sequentially composed, and what things need to be *interleaved*. `flatMap` on `Stream` is for interleaving, while `++` is for sequential composition. When the `Stream` is a singleton, there is no distinction between the two other than resource scoping. When the `Stream` is *not* a singleton, the difference is of course massive.

I actually didn't check the S3 and Postgres changes sufficiently closely to rule out this problem; I'm glad that ThoughtSpot failed. We need to be very careful of this. The good news is that this complexity is inherent to the space *regardless* of the `ResultSink` return type, and it's actually a big part of the reason why returning `Stream` is the correct thing to do.